### PR TITLE
Add RNA modification and fusion detection

### PR DESCRIPTION
**Pull request by yuukiiwa**
Add RNA modification and fusion detection functionalities to `nanoseq`
## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
    - [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/nanoseq/tree/master/.github/CONTRIBUTING.md)
    - [x] If necessary, also make a PR on the nf-core/nanoseq _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: subworkflows/local/rna_fusions_jaffal.nf [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    ch_sample
+        .map { it -> [ it[0], it[6] ]}
+        .set { ch_jaffal_input }
+
+    /*
+    * Align current signals to reference with Nanopolish
+    */
+    JAFFAL( ch_jaffal_input, ch_jaffal_ref_dir )
+
+//    emit:
```
Comment (christopher-hakkaart) [inline subworkflows/local/rna_fusions_jaffal.nf:30]: I don't know much about this analysis but I just wanted to check these were commented on purpose

File: subworkflows/local/rna_fusions_jaffal.nf [hunk 2]
```diff
@@ -0,0 +1,34 @@
+include { GET_JAFFAL_REF } from '../../modules/local/get_jaffal_ref'
```
Comment (christopher-hakkaart) [inline subworkflows/local/rna_fusions_jaffal.nf:1]: Can you add a heading here?

File: workflows/nanoseq.nf [hunk 3]
```diff
@@ -336,6 +349,9 @@ workflow NANOSEQ{
         ch_view_sortbam
             .map { it -> [ it[0], it[3] ] }
             .set { ch_sortbam }
+        ch_view_sortbam
+            .map { it -> [ it[0], it[3], it[4] ] }
+            .set { ch_nanopolish_sortbam }
```
Comment (christopher-hakkaart) [inline workflows/nanoseq.nf:354]: Are both of these needed or can it be made into an if statement?

File: modules/local/m6anet_inference.nf [hunk 4]
```diff
@@ -0,0 +1,26 @@
+process M6ANET_INFERENCE {
+    tag "$meta.id"
+    echo true
+    label 'process_medium'
+
+//  conda     (params.enable_conda ? "bioconda::nanopolish==0.13.2" : null) // need to get xpore onto conda
```
Comment (christopher-hakkaart) [inline modules/local/m6anet_inference.nf:6]: No conda or singularity?

File: modules/local/m6anet_dataprep.nf [hunk 5]
```diff
@@ -0,0 +1,21 @@
+process M6ANET_DATAPREP {
+    tag "$meta.id"
+    label 'process_medium'
+
+//  conda     (params.enable_conda ? "bioconda::nanopolish==0.13.2" : null) // need to get xpore onto conda
```
Comment (christopher-hakkaart) [inline modules/local/m6anet_dataprep.nf:5]: No conda or singularity?

File: modules/local/m6anet_dataprep.nf [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    output:
+    tuple val(meta), path("$meta.id"), emit: dataprep_outputs
+
+    script:
+    """
+    m6anet-dataprep \\
+    --eventalign $eventalign \\
+    --out_dir $meta.id \\
+    --n_processes $task.cpus
+    """
```
Comment (christopher-hakkaart) [inline modules/local/m6anet_dataprep.nf:20]: Can process and version information be added?

File: modules/local/xpore_dataprep.nf [hunk 7]
```diff
@@ -0,0 +1,22 @@
+process XPORE_DATAPREP {
+    tag "$meta.id"
+    label 'process_medium'
+
+//  conda     (params.enable_conda ? "bioconda::nanopolish==0.13.2" : null) // need to get xpore onto conda
```
Comment (christopher-hakkaart) [inline modules/local/xpore_dataprep.nf:5]: Conda or singularity?

File: modules/local/xpore_dataprep.nf [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    tuple val(meta), path("$meta.id"), emit: dataprep_outputs
+
+    script:
+    """
+    xpore dataprep \\
+    --eventalign $eventalign \\
+    --out_dir $meta.id \\
+    --n_processes $task.cpus \\
+    --genome --gtf_or_gff $gtf --transcript_fasta $genome
+    """
```
Comment (christopher-hakkaart) [inline modules/local/xpore_dataprep.nf:21]: Can version and process information be added?

File: modules/local/jaffal.nf [hunk 9]
```diff
@@ -0,0 +1,23 @@
+process JAFFAL {
+    tag "$meta.id"
+    echo true
+    label 'process_medium'
+
+//    conda     (params.enable_conda ? "bioconda::nanolyse=1.2.0" : null)   //need conda container
```
Comment (christopher-hakkaart) [inline modules/local/jaffal.nf:6]: No conda or singularity?

File: modules/local/jaffal.nf [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+    output:
+    tuple val(meta), path("*.fasta") ,emit: jaffal_fastq
+    path "*.csv"                     ,emit: jaffal_results
+    path "*_version.txt"             ,emit: version
+
+    script:
+    """
+    bpipe run -p refBase=$jaffal_ref_dir $jaffal_ref_dir/JAFFAL.groovy $fastq
+    echo '2.0' > jaffal_version.txt
```
Comment (christopher-hakkaart) [inline modules/local/jaffal.nf:21]: 'jaffa 2.0'

**Comment by yuukiiwa** [comment]
Happy New Year! Thank you for the review, @christopher-hakkaart! I have merged a fix needed for the pipeline to run on `Azure`, which is not quite compatible with most of the functionalities. I will try to fix these starting on the week of January 24th, and in the meanwhile, I am submitting some `Dockerfile`s to `Biocontainers` to address your comment on the missing `singularity` containers for some tools, and the reviewing processes on `Biocontainers` will take quite a bit of time.
Thanks!

**Comment by christopher-hakkaart** [comment]
Hi @yuukiiwa ! Sorry I've been away on the last of my holidays. Overall it all looks really good and the new functionality is really cool. I do think it's worthwhile to add containers before merging to keep the dev branch as complete as possible. Once the containers are up I can approve the extra changes for a quick merge. If you want a merge before that let me know and we can sort something out.

**Comment by yuukiiwa** [comment]
Thank you so much, @christopher-hakkaart! I will let you know once the PR is ready!

File: subworkflows/local/rna_modifications_xpore_m6anet.nf [hunk 11]
```diff
@@ -0,0 +1,60 @@
+include { NANOPOLISH_INDEX_EVENTALIGN } from '../../modules/local/nanopolish_index_eventalign'
```
[the remark this answers is not in the document]
Comment (yuukiiwa) [reply subworkflows/local/rna_modifications_xpore_m6anet.nf:1]: Yes, I have added a heading here

File: workflows/nanoseq.nf (hunk 3 above)
Comment (yuukiiwa) [reply workflows/nanoseq.nf:361]: Yes, both of these are needed for the RNA track: `ch_sortbam` is for bigbed, bigwig, and RNA quantification, and RNA fusion detection while `ch_nanopolish_sortbam` is for RNA modification detection.

File: subworkflows/local/rna_fusions_jaffal.nf (hunk 2 above)
Comment (yuukiiwa) [reply subworkflows/local/rna_fusions_jaffal.nf:1]: Yes, I have added a heading here

File: modules/local/xpore_dataprep.nf (hunk 7 above)
Comment (yuukiiwa) [reply modules/local/xpore_dataprep.nf:5]: Yes, I have added conda and singularity

File: modules/local/jaffal.nf (hunk 9 above)
Comment (yuukiiwa) [reply modules/local/jaffal.nf:6]: Yes, I have added conda for now. I am contacting the developers for JAFFAL to update their biocontainer, which is not working right now

File: modules/local/jaffal.nf (hunk 10 above)
Comment (yuukiiwa) [reply modules/local/jaffal.nf:21]: Yes, I have changed this to `jaffa 2.0`

File: modules/local/m6anet_inference.nf (hunk 4 above)
Comment (yuukiiwa) [reply modules/local/m6anet_inference.nf:6]: Our lab is trying to get `m6anet` onto `bioconda`, but the person who wrote `m6anet` is very busy right now, so I guess we will include `conda` and `singularity` in a future release instead.

File: modules/local/m6anet_dataprep.nf (hunk 5 above)
Comment (yuukiiwa) [reply modules/local/m6anet_dataprep.nf:5]: Please see comment above

File: subworkflows/local/rna_fusions_jaffal.nf (hunk 1 above)
Comment (yuukiiwa) [reply subworkflows/local/rna_fusions_jaffal.nf:30]: I forgot to clean up. These are not related to the RNA fusion detection analysis

File: modules/local/m6anet_dataprep.nf (hunk 6 above)
Comment (yuukiiwa) [reply modules/local/m6anet_dataprep.nf:20]: Yes, I have added that

File: modules/local/xpore_dataprep.nf (hunk 8 above)
Comment (yuukiiwa) [reply modules/local/xpore_dataprep.nf:23]: Yes, I have added that

**Comment by yuukiiwa** [comment]
Hi @christopher-hakkaart, I have updated the PR which is now compatible with the Azure fix!

**Review by christopher-hakkaart — APPROVED** [review]
Thanks for addressing my comments. I didn't spot anything that looked out of place. Everything looks really good.

**Pull request merged**